xarray/backends/api.py

```diff
@@ -532,68 +538,39 @@ def maybe_decode_store(store, chunks, lock=False):
 
     if isinstance(filename_or_obj, AbstractDataStore):
         store = filename_or_obj
+    else:
+        if isinstance(filename_or_obj, str):
+            filename_or_obj = _normalize_path(filename_or_obj)
+
+            if engine is None:
+                engine = _get_default_engine(filename_or_obj, allow_remote=True)
+        elif isinstance(filename_or_obj, MutableMapping) and engine == "zarr":
```

Comment:
I think this was a bug introduced in #4187. Not exactly sure what we want but we used to support zarr filenames as strings. This line now requires they be a `MutableMapping`. Curious to hear from others but we may be able to get away with the following.

```suggestion
        elif engine == "zarr":
```

Reply:
cc @weiji14 on this one!

Reply:
See https://github.com/intake/intake-xarray/issues/70#issuecomment-621268486 for context, if I'm reading it correctly, it's so that remote Zarr files can be accessed via [`fsspec.get_mapper`](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.get_mapper).

Reply:
@weiji14 - The problem with the current implementation is that `open_dataset("./filename.zarr", engine="zarr")` will raise the `ValueError("can only read bytes...")` before getting down to lines 555:557.

To illustrate further, we caught this in our nightly upstream test on xpublish: https://github.com/xarray-contrib/xpublish/runs/1156500330?check_suite_focus=true

Reply:
@jhamman the `if isinstance(filename_or_obj, str):` at line 542 handles the `str` case for 'zarr' before the `elif`. No?

Reply:
I confirm, in this PR paths passed as strings reach line 573 as expected:

`store = opener(filename_or_obj, **extra_kwargs, **backend_kwargs)`

Reply:
@jhamman, if I'm reading the logs correctly, `filename_or_obj` is a `tests.utils.TestMapper` object which inherits from `starlette.testclient.TestClient`. I think the tests might pass again if we use `xarray.core.utils.is_dict_like(...)` instead of `isinstance(..., MutableMapping)`

Reply:
Can we avoid all type checks on `filename_or_obj` when engine is explicitly provided? It seems like we could safely just pass it on to the datastore object?

Reply:
I just realized we've been talking about the changes in current `master`, and most of this is already fixed by this PR. If it makes sense to always set `mode="r"` I guess we can simply remove the `isinstance` check.

Reply:
@shoyer 

> Can we avoid all type checks on `filename_or_obj` when engine is explicitly provided? It seems like we could safely just pass it on to the datastore object?

Not sure. Most type checks conditionals apply generic transformations, like conversion from `Path` to `str` and `_normalize_path` that are probably needed by all backends.

Also, the logic for the selection of the default engine varies depending on the `filename_or_obj` data type.

I only see a possible simplification in the handling of the opened-file case but at the cost of changing the current behaviour (in `master` and in this PR if you attempt to open a netCDF3 with `engine='h5netcdf'` you end up silently using `engine='scipy'`).

I would suggest to make a different PR for that change.

Reply:
I can add this change that makes `mode="r"` unconditional for 'zarr' and cleans up a bit the ifs (Note sure how to show you without making the commit):

```diff
Index: xarray/backends/api.py
I===================================================================
--- xarray/backends/api.py	(revision 4a1319243692e2a4425f9a7faf0c5df8a80a630d)
+++ xarray/backends/api.py	(date 1600901329932)
@@ -544,12 +543,7 @@
 
             if engine is None:
                 engine = _get_default_engine(filename_or_obj, allow_remote=True)
-        elif isinstance(filename_or_obj, MutableMapping) and engine == "zarr":
-            # Zarr supports a wide range of access modes, but for now xarray either
-            # reads or writes from a store, never both.
-            # For open_dataset(engine="zarr"), we only read (i.e. mode="r")
-            extra_kwargs["mode"] = "r"
-        else:
+        elif engine != "zarr":
             if engine not in [None, "scipy", "h5netcdf"]:
                 raise ValueError(
                     "can only read bytes or file-like objects "
@@ -567,6 +561,7 @@
             overwrite_encoded_chunks = backend_kwargs.pop(
                 "overwrite_encoded_chunks", None
             )
+            extra_kwargs["mode"] = "r"
             extra_kwargs["group"] = group
 
         opener = _get_backend_cls(engine)
```

Reply:
> I can add this change...

Done